text/3451-additional-float-types.md

```diff
@@ -0,0 +1,139 @@
+
+# Motivation
+[motivation]: #motivation
+
+IEEE-754 standard defines binary floating point formats, including binary16, binary32, binary64 and binary128. The binary32 and binary64 correspond to `f32` and `f64` types in Rust, while binary16 and binary128 are used in multiple scenarios (machine learning, scientific computing, etc.) and accepted by some modern architectures (by software or hardware).
+
+In C/C++ world, there're already types representing these formats, along with more legacy non-standard types specific to some platform. Introduce them in a limited way would help improve FFI against such code.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+`f16` and `f128` are primitive floating types, they can be used just like `f32` or `f64`. They always conform to binary16 and binary128 format defined in IEEE-754, which means size of `f16` is always 16-bit, and size of `f128` is always 128-bit.
+
+```rust
+let val1 = 1.0; // Default type is still f64
+let val2: f128 = 1.0;
+let val3: f16 = 1.0;
+let val4 = 1.0f128; // Suffix of f128 literal
+let val5 = 1.0f16; // Suffix of f16 literal
+
+println!("Size of f128 in bytes: {}", std::mem::size_of_val(&val2)); // 16
+println!("Size of f16 in bytes: {}", std::mem::size_of_val(&val3)); // 2
+```
+
+Because not every target supports `f16` and `f128`, compiler provides conditional guards for them:
+
+```rust
+#[cfg(target_has_f128)]
+fn get_f128() -> f128 { 1.0f128 }
+
+#[cfg(target_has_f16)]
+fn get_f16() -> f16 { 1.0f16 }
+```
+
+All operators, constants and math functions defined for `f32` and `f64` in core, are also defined for `f16` and `f128`, and guarded by respective conditional guards.
+
+`f80` type is defined in `core::arch::{x86, x86_64}`. `doubledouble` type is defined in `core::arch::{powerpc, powerpc64}`. `bf16` type is defined in `core::arch::{arm, aarch64, x86, x86_64}`. They do not have literal representation.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `f16` type
+
+`f16` consists of 1 bit of sign, 5 bits of exponent, 10 bits of mantissa.
+
+The following `From` and `TryFrom` traits are implemented for conversion between `f16` and other types:
+
+```rust
+impl From<f16> for f32 { /* ... */ }
+impl From<f16> for f64 { /* ... */ }
+impl From<bool> for f16 { /* ... */ }
+impl From<u8> for f16 { /* ... */ }
+impl From<i8> for f16 { /* ... */ }
+```
+
+`f16` will generate `half` type in LLVM IR.
+
+## `f128` type
+
+`f128` consists of 1 bit of sign, 15 bits of exponent, 112 bits of mantissa.
+
+`f128` is available for on targets having (1) hardware instructions or software emulation for 128-bit float type; (2) backend support for `f128` type on the target; (3) essential target features enabled (if any).
+
+The list of targets supporting `f128` type may change over time. Initially, it includes `powerpc64le-*`.
```

Comment:
```suggestion
The list of targets supporting `f128` type may change over time. Initially, it includes `powerpc64le-*`.
`x86_64-*` and `aarch64-*`
```

I don't know for sure what targets support it, but we should aim to at least support the major 64-bit CPUs
here at first